solutions/go/ping-pong-multiple/code/app/server.go

```diff
@@ -0,0 +1,30 @@
+package main
+
+import (
+	"fmt"
+	"net"
+	"os"
+)
+
+func main() {
+	l, err := net.Listen("tcp", "0.0.0.0:6379")
+	if err != nil {
+		fmt.Println("Failed to bind to port 6379")
+		os.Exit(1)
+	}
+
+	conn, err := l.Accept()
+	if err != nil {
+		fmt.Println("Error accepting connection: ", err.Error())
+		os.Exit(1)
+	}
+
+	for {
+		if _, err := conn.Read([]byte{}); err != nil {
```

Comment:
How are you testing this? This solution is not correct, since this `conn.Read` doesn't give any guarantees that it'll be parsing the commands independently. The correct way to implement this is: 

```go
package main

import (
	"bufio"
	"fmt"

	// Uncomment this block to pass the first stage
	"net"
	"os"
)

func main() {
	// You can use print statements as follows for debugging, they'll be visible when running tests.
	fmt.Println("Logs from your program will appear here!")

	// Uncomment this block to pass the first stage
	//
	l, err := net.Listen("tcp", "0.0.0.0:6379")
	if err != nil {
		fmt.Println("Failed to bind to port 6379")
		os.Exit(1)
	}

	c, err := l.Accept()
	if err != nil {
		fmt.Println("Error accepting connection: ", err.Error())
		os.Exit(1)
	}
        defer c.Close()
	scanner := bufio.NewScanner(c)

	for scanner.Scan() {
		c.Write([]byte("+PONG\r\n"))
	}
}
```

Reply:
@marcosnils all solutions are tested [using the official Redis client](https://github.com/codecrafters-io/redis-tester/blob/da9590868d33fa7351409fd3b8e13340c8953cb9/internal/test_ping_pong.go#L35-L54), but you're right - `Read()` could get called multiple times for a single command if all input isn't returned at once. I think the reason these tests pass is that they're only looking for a response (irrespective of whether that response is sent _after_ the command is fully parsed or not).

I'm not sure if `bufio.NewScanner` solves this problem though... An echo command is sent as an array (`["echo", "hey"]`), which will look something like this when encoded: `*2\r\n$4\r\necho\r\n$3\r\nhey\r\n`. Since `bufio.NewScanner` uses [`ScanLines`](https://pkg.go.dev/bufio#ScanLines) by default, the first `Scan()` will split at `\r\n` and just return `*2`, no? 

I can't think of an easy way to ensure we parse a whole "command" without introducing a proper RESP parser like we do in stage 5 (when the user needs to to parse commands):
 
<img width="947" alt="Screenshot 2022-07-29 at 12 28 00" src="https://user-images.githubusercontent.com/3893573/181682922-134f2662-0071-4c1a-ad1f-140a979c2433.png">

@marcosnils I think the common approach we'd follow here is to add a note in the explanation saying that "this is a problem, but we'll tackle it in stage 5". I have a PR for this here: https://github.com/codecrafters-io/build-your-own-redis/pull/36/files.

Reply:
> I'm not sure if `bufio.NewScanner` solves this problem though... An echo command is sent as an array (`["echo", "hey"]`), which will look something like this when encoded: `*2\r\n$4\r\necho\r\n$3\r\nhey\r\n`. Since `bufio.NewScanner` uses [`ScanLines`](https://pkg.go.dev/bufio#ScanLines) by default, the first `Scan()` will split at `\r\n` and just return `*2`, no?

this is not the case for PING since it doesn't have any arguments. Therefore, in the write, the PING <> PONG exchange looks like: 

```
PING\r\n
+PONG
PING\r\n
+PONG
PING\r\n
+PONG
```

IMO it's still useful to introduce the `scanner` at this stance since you could also print the `PING` command to make it more realistic to users.

Reply:
@marcosnils the format you're referring to above is the ["inline commands"](https://redis.io/docs/reference/protocol-spec/#inline-commands) format, yes? Scanner would work well for that, yeah. The official Redis client we use in tests doesn't use the inline commands format though, it sends ping like this: `*1\r\n$4\r\nping\r\n` (i.e. `["ping"]` encoded as a RESP array). From what I can tell, it uses this format irrespective of whether a command has arguments or not.

Reply:
> @marcosnils the format you're referring to above is the ["inline commands"](https://redis.io/docs/reference/protocol-spec/#inline-commands) format, yes? Scanner would work well for that, yeah. The official Redis client we use in tests doesn't use the inline commands format though, it sends ping like this: `*1\r\n$4\r\nping\r\n` (i.e. `["ping"]` encoded as a RESP array). From what I can tell, it uses this format irrespective of whether a command has arguments or not.

that's correct. Completely forgot about inline format. 

In that case, I think it's worth to change the `Read` part of the exercise to at least write `+PONG\r\n` when a `ping` is found. This could easily be achieved by writing a custom `split` function for the scanner, or just counting the number of `ping` occurrences in the read bytes.

I think doing a `strings.Count(<input>, "ping")` and reply based on that should be pretty easy and serves as a starting point to let the users understand the redis protocol. 

WDYT?

Reply:
> I think doing a strings.Count(<input>, "ping") and reply based on that should be pretty easy and serves as a starting point to let the users understand the redis protocol.

@marcosnils yep, agreed! I'll make changes to both stage 2 & 3 and update this PR - will tag you once done 👍

Reply:
We've changed the format of these repositories a bit, I've opened a PR agains the new format here: https://github.com/codecrafters-io/build-your-own-redis/pull/49. Will tag there!